File: BAEKJOON/그래프 탐색/바이러스/바이러스.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <set>

using namespace std;

set<int> bfs(const vector<vector<int>>& graph, int start) {
    set<int> visited;
    queue<int> q;
    q.push(start);
    visited.insert(start);
    
    while (!q.empty()) {
        int v = q.front();
        q.pop();
        for (int neighbor : graph[v]) {
            if (visited.find(neighbor) == visited.end()) {
                visited.insert(neighbor);
                q.push(neighbor);
            }
        }
    }
    return visited;
}

int main() {
    int n, m;
    cin >> n >> m;
    
    vector<vector<int>> graph(n + 1); // 컴퓨터 번호가 1부터 시작하므로 n+1 크기
    
    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }
    
    set<int> infected_computers = bfs(graph, 1);
    
    // 1번 컴퓨터를 제외
    infected_computers.erase(1);
    
    // 결과 출력
    cout << infected_computers.size() << endl;
    
    return 0;
}

File: BAEKJOON/그래프 탐색/바이러스/바이러스.cs
using System;
using System.Collections.Generic;

class Program
{
    static HashSet<int> Bfs(Dictionary<int, List<int>> graph, int start)
    {
        HashSet<int> visited = new HashSet<int>();
        Queue<int> queue = new Queue<int>();
        queue.Enqueue(start);
        visited.Add(start);
        
        while (queue.Count > 0)
        {
            int v = queue.Dequeue();
            foreach (int neighbor in graph[v])
            {
                if (!visited.Contains(neighbor))
                {
                    visited.Add(neighbor);
                    queue.Enqueue(neighbor);
                }
            }
        }
        return visited;
    }

    static void Main()
    {
        int n = int.Parse(Console.ReadLine().Trim());  // 컴퓨터의 수
        int m = int.Parse(Console.ReadLine().Trim());  // 네트워크 상에서 직접 연결되어 있는 컴퓨터 쌍의 수

        // 그래프 초기화
        Dictionary<int, List<int>> graph = new Dictionary<int, List<int>>();
        for (int i = 1; i <= n; i++)
        {
            graph[i] = new List<int>();
        }

        // 간선 입력 받기
        for (int i = 0; i < m; i++)
        {
            string[] inputs = Console.ReadLine().Trim().Split();
            int a = int.Parse(inputs[0]);
            int b = int.Parse(inputs[1]);
            graph[a].Add(b);
            graph[b].Add(a);
        }

        // BFS를 통해 1번 컴퓨터와 연결된 모든 컴퓨터를 찾기
        HashSet<int> infectedComputers = Bfs(graph, 1);

        // 1번 컴퓨터 제외
        infectedComputers.Remove(1);

        // 결과 출력
        Console.WriteLine(infectedComputers.Count);
    }
}
